updater/updater.go

```diff
@@ -0,0 +1,249 @@
+			Result: omaha.EventResultSuccess,
+		}
+	case ProgressUpdateCompleteAndRestarted:
+		return &omaha.EventRequest{
+			Type:   omaha.EventTypeUpdateComplete,
+			Result: omaha.EventResultSuccessReboot,
+		}
+	case ProgressInstallationStarted:
+		return &omaha.EventRequest{
+			Type:   omaha.EventTypeInstallStarted,
+			Result: omaha.EventResultSuccess,
+		}
+	case ProgressInstallationFinished:
+		return &omaha.EventRequest{
+			Type:   omaha.EventTypeInstallStarted,
+			Result: omaha.EventResultSuccess,
+		}
+	case ProgressError:
+		return &omaha.EventRequest{
+			Type:   omaha.EventTypeUpdateComplete,
+			Result: omaha.EventResultError,
+		}
+	default:
+		return nil
+	}
+}
+
+type OmahaRequestHandler interface {
+	Handle(req *omaha.Request) (*omaha.Response, error)
+}
+
+type Updater struct {
+	omahaURL      string
+	clientVersion string
+
+	instanceID      string
+	instanceVersion string
+	sessionID       string
+
+	appID   string
+	channel string
+
+	debug           bool
+	omahaReqHandler OmahaRequestHandler
+
+	mu sync.RWMutex
+}
+
+type Config struct {
+	OmahaURL        string
+	AppID           string
+	Channel         string
+	InstanceID      string
+	InstanceVersion string
+	Debug           bool
+	OmahaReqHandler OmahaRequestHandler
+}
+
+func New(config Config) (*Updater, error) {
+	if _, err := url.Parse(config.OmahaURL); err != nil {
+		return nil, fmt.Errorf("parsing URL %q: %w", config.OmahaURL, err)
+	}
+	updater := Updater{
+		omahaURL:        config.OmahaURL,
+		clientVersion:   defaultClientVersion,
+		instanceID:      config.InstanceID,
+		sessionID:       uuid.New().String(),
+		appID:           config.AppID,
+		instanceVersion: config.InstanceVersion,
+		channel:         config.Channel,
+		debug:           config.Debug,
+	}
+	if config.OmahaReqHandler == nil {
+		updater.omahaReqHandler = NewDefaultOmahaRequestHandler(config.OmahaURL)
+	} else {
+		updater.omahaReqHandler = config.OmahaReqHandler
+	}
+	return &updater, nil
+}
+
+func NewAppRequest(u *Updater) *omaha.Request {
+	req := omaha.NewRequest()
+	req.Version = u.clientVersion
+	req.UserID = u.instanceID
+	req.SessionID = u.sessionID
+
+	app := req.AddApp(u.appID, u.GetInstanceVersion())
+	app.MachineID = u.instanceID
+	app.BootID = u.sessionID
+	app.Track = u.channel
+
+	return req
+}
+
+func (u *Updater) SendOmahaRequest(req *omaha.Request) (*omaha.Response, error) {
+	if u.debug {
+		requestByte, err := xml.Marshal(req)
+		if err == nil {
+			fmt.Println("Raw Request:\n", string(requestByte))
+		}
+	}
+	resp, err := u.omahaReqHandler.Handle(req)
+	if u.debug {
+		responseByte, err := xml.Marshal(resp)
+		if err == nil {
+			fmt.Println("Raw Response:\n", string(responseByte))
+		}
+	}
+	return resp, err
+}
+
+func (u *Updater) CheckForUpdates(ctx context.Context) (*UpdateInfo, error) {
+	req := NewAppRequest(u)
+	app := req.GetApp(u.appID)
+	app.AddUpdateCheck()
+
+	resp, err := u.SendOmahaRequest(req)
+	if err != nil {
+		return nil, err
```

Comment:
Errors should be annotated:
```suggestion
		return nil, fmt.Errorf("sending omaha request: %w", err)
```